File: tools/message_side_channel.py
"""
Simple Python-side implementation of custom Unity-Python side channels that
communicate string messages aside the main mlagents pipeline
"""
from mlagents_envs.side_channel.side_channel import (
    SideChannel,
    OutgoingMessage,
)
import uuid


class StringMsgChannel(SideChannel):
    def __init__(self, channel_uuid):
        super().__init__(uuid.UUID(channel_uuid))
        self.incoming_message_buffer = []

    def on_message_received(self, msg):
        # Read until end of message

        # Speaker info & utterance content as string
        speaker = msg.read_string()
        utterance = msg.read_string()

        # Retrieve any demonstrative references (map from substring indices to
        # segmentation masks on visual observation image) until end of the
        # current message is reached (signalled by -1)
        dem_refs = {}
        while True:
            int_message = msg.read_int32()

            if int_message == -1:
                # End of message
                break
            else:
                # Read integer is substring start index; process the remainder
                start = int_message
                end = msg.read_int32()
                ref_by_mask = msg.read_bool()
                if ref_by_mask:
                    # Reference by segmentation mask (list of floats in 0~1)
                    dem_refs[(start, end)] = self.rle_decode(msg.read_float32_list())
                else:
                    # Reference by string name
                    dem_refs[(start, end)] = msg.read_string()

        # Put processed message data into incoming buffer
        self.incoming_message_buffer.append((speaker, utterance, dem_refs))

    def send_string(self, speaker, utterance, dem_refs):
        assert isinstance(utterance, str) and isinstance(dem_refs, dict)

        msg = OutgoingMessage()

        # Utterance content as string
        msg.write_string(speaker)
        msg.write_string(utterance)

        # Write any demonstrative references
        for (start, end), ref in dem_refs.items():
            msg.write_int32(start); msg.write_int32(end)
            if isinstance(ref, list):
                # Reference by segmentation mask (list of floats in 0~1)
                assert all(x>=0 and x<=1 for x in ref)
                msg.write_bool(True)
                msg.write_float32_list(self.rle_encode(ref))
            else:
                # Reference by string name (of Unity GameObject)
                assert isinstance(ref, str)
                msg.write_bool(False)
                msg.write_string(ref)

        # Mark end of message segment
        msg.write_int32(-1)

        # Queue message to send
        super().queue_message_to_send(msg)

    @staticmethod
    def rle_encode(raw_mask):
        # Encode raw binary mask into RLE format for message compression
        rle = []

        zero_flag = True
        run = 0
        for f in raw_mask:
            # Increment run length if value matches with current flag; push current
            # run length value to return array and flip sign
            if zero_flag:
                if f == 0: run += 1
                else:
                    rle.append(run)
                    zero_flag = False
                    run = 1
            else:
                if f > 0: run += 1
                else:
                    rle.append(run)
                    zero_flag = True
                    run = 1
        if run > 0: rle.append(run)     # Flush last entry

        return rle

    @staticmethod
    def rle_decode(rle_mask):
        # Decode RLE to recover raw binary mask
        total_length = int(sum(rle_mask))
        raw = [None] * total_length

        zero_flag = True
        cumulative = 0
        for f in rle_mask:
            # Get integer run length and update values
            run_length = int(f)
            for i in range(cumulative, cumulative+run_length):
                raw[i] = 0 if zero_flag else 1
            
            # Flip sign and update cumulative index
            zero_flag = not zero_flag
            cumulative += run_length

        return raw

File: unity/Assets/Scripts/MessageSideChannel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Unity.MLAgents.SideChannels;

public class MessageSideChannel : SideChannel
{
    // Associated agent in scene
    private readonly DialogueAgent _listeningAgent;

    public MessageSideChannel(string channelUuid, DialogueAgent agent)
    {
        ChannelId = new Guid(channelUuid);
        _listeningAgent = agent;
    }

    protected override void OnMessageReceived(IncomingMessage msg)
    {
        // Read until end of message
        var speaker = msg.ReadString();
        var utterance = msg.ReadString();

        // Retrieve any demonstrative references (map from substring indices to
        // segmentation mask image, having same dimension as scene image perceived
        // by sensor) until end of the current message is reached (signalled by -1)
        var demRefs = new Dictionary<(int, int), EntityRef>();
        while (true)
        {
            var intMessage = msg.ReadInt32();
        
            if (intMessage == -1) break;        // End of dem. refs list
            
            var start = intMessage;
            var end = msg.ReadInt32();
            var refByMask = msg.ReadBoolean();
            if (refByMask)
            {
                var rleMask = msg.ReadFloatList().ToArray();
                demRefs[(start, end)] = new EntityRef(RleDecode(rleMask));
            }
            else
                demRefs[(start, end)] = new EntityRef(msg.ReadString());
        }

        // Handle system requests from backend (only concerns ground-truth masks info
        // requests currently)
        if (speaker == "System" && utterance.StartsWith("GT mask request: "))
        {
            var requests = utterance.Replace("GT mask request: ", "");
            foreach (var req in requests.Split(", "))
                _listeningAgent.gtMaskRequests.Enqueue(req);
        }

        // Put processed message data into incoming buffer
        if (speaker == _listeningAgent.dialogueParticipantID)
            _listeningAgent.outgoingMsgBuffer.Enqueue((speaker, utterance, demRefs));
    }

    public void SendMessageToBackend(
        string speaker, string utterance, Dictionary<(int, int), EntityRef> demRefs
    )
    {
        // Create OutgoingMessage instance (using for dispose at the end)
        using var msgOut = new OutgoingMessage();

        // Write speaker info & utterance content as string
        msgOut.WriteString(speaker);
        msgOut.WriteString(utterance);

        // (If any) Encode demonstrative references as two consecutive ints (marking
        // start & end of corresponding demonstrative pronoun substring) and either
        // float[] ((soft) segmentation mask) or string (direct reference by string
        // name of EnvEntity)
        foreach (var (range, demRef) in demRefs)
        {
            var (start, end) = range;
            msgOut.WriteInt32(start); msgOut.WriteInt32(end);
            switch (demRef.refType)
            {
                case EntityRefType.Mask:
                    msgOut.WriteBoolean(true);
                    msgOut.WriteFloatList(RleEncode(demRef.maskRef));
                    break;
                case EntityRefType.String:
                    msgOut.WriteBoolean(false);
                    msgOut.WriteString(demRef.stringRef);
                    break;
                default:
                    // Shouldn't reach here but anyways
                    throw new Exception("Invalid reference data type?");
            }
        }

        // Mark end of message segment
        msgOut.WriteInt32(-1);

        // Queue message to send
        QueueMessageToSend(msgOut);
    }

    private static float[] RleEncode(float[] rawMask)
    {
        // Encode raw binary mask into RLE format for message compression
        var rle = new List<float>();

        var zeroFlag = true;
        var run = 0f;
        foreach (var f in rawMask)
        {
            // Increment run length if value matches with current flag; push current
            // run length value to return array and flip sign
            if (zeroFlag)
            {
                if (f == 0f) run += 1;
                else
                {
                    rle.Add(run);
                    zeroFlag = false;
                    run = 1f;
                }
            }
            else
            {
                if (f > 0f) run += 1;
                else
                {
                    rle.Add(run);
                    zeroFlag = true;
                    run = 1f;
                }
            }
        }
        if (run > 0f) rle.Add(run);     // Flush last entry

        return rle.ToArray();
    }

    private static float[] RleDecode(float[] rleMask)
    {
        // Decode RLE to recover raw binary mask
        var totalLength = (int) rleMask.Sum();
        var raw = new float[totalLength];

        var zeroFlag = true;
        var cumulative = 0; 
        foreach (var f in rleMask)
        {
            // Get integer run length and update values 
            var runLength = (int) f;
            for (var i = cumulative; i < cumulative+runLength; i++)
                raw[i] = zeroFlag ? 0f : 1f;

            // Flip sign and update cumulative index
            zeroFlag = !zeroFlag;
            cumulative += runLength;
        }

        return raw;
    }
}
